src/rust/cryptography-x509-validation/src/policy/mod.rs

```diff
@@ -250,6 +285,277 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
+                // 5280 4.2.1.9: Basic Constraints
+                ExtensionPolicy::present(
+                    BASIC_CONSTRAINTS_OID,
+                    Criticality::Critical,
+                    Some(ca::basic_constraints),
+                ),
+                // 5280 4.2.1.10: Name Constraints
+                ExtensionPolicy::maybe_present(NAME_CONSTRAINTS_OID, Criticality::Critical, None),
+                // 5280 4.2.1.10: Policy Constraints
+                ExtensionPolicy::maybe_present(POLICY_CONSTRAINTS_OID, Criticality::Critical, None),
+            ]),
+            ee_extension_policies: Vec::from([
+                // 5280 4.2.1.1.: Authority Key Identifier
+                ExtensionPolicy::present(
+                    AUTHORITY_KEY_IDENTIFIER_OID,
+                    Criticality::NonCritical,
+                    None,
+                ),
+                // 5280 4.2.1.3: Key Usage
+                ExtensionPolicy::maybe_present(KEY_USAGE_OID, Criticality::Agnostic, None),
+                // CA/B 7.1.2.7.12 Subscriber Certificate Subject Alternative Name
+                ExtensionPolicy::present(
+                    SUBJECT_ALTERNATIVE_NAME_OID,
+                    Criticality::Agnostic,
+                    Some(ee::subject_alternative_name),
+                ),
+                // 5280 4.2.1.9: Basic Constraints
+                ExtensionPolicy::maybe_present(
+                    BASIC_CONSTRAINTS_OID,
+                    Criticality::Agnostic,
+                    Some(ee::basic_constraints),
+                ),
+                // 5280 4.2.1.10: Name Constraints
+                ExtensionPolicy::not_present(NAME_CONSTRAINTS_OID),
+                // CA/B 7.1.2.7.10 Subscriber Certificate Extended Key Usage
+                ExtensionPolicy::present(
+                    EXTENDED_KEY_USAGE_OID,
+                    Criticality::NonCritical,
+                    Some(ee::extended_key_usage),
+                ),
+            ]),
+        }
+    }
+
+    fn permits_basic(&self, cert: &Certificate<'_>) -> Result<(), PolicyError> {
+        let extensions = cert.extensions()?;
+
+        // CA/B 7.1.1:
+        // Certificates MUST be of type X.509 v3.
+        if cert.tbs_cert.version != 2 {
+            return Err("certificate must be an X509v3 certificate".into());
+        }
+
+        // 5280 4.1.1.2 / 4.1.2.3: signatureAlgorithm / TBS Certificate Signature
+        // The top-level signatureAlgorithm and TBSCert signature algorithm
+        // MUST match.
+        if cert.signature_alg != cert.tbs_cert.signature_alg {
+            return Err("mismatch between signatureAlgorithm and SPKI algorithm".into());
+        }
+
+        // 5280 4.1.2.2: Serial Number
+        let serial_bytes = cert.tbs_cert.serial.as_bytes();
+        if serial_bytes.len() == 1 && serial_bytes[0] == 0 {
+            // The serial number MUST be a positive integer.
+            return Err("certificate serial number must not be 0".into());
```

Comment:
This ignores the case where the serial is negative.

Reply:
Whoops, thanks for catching that!

Reply:
0c5ff83

Will follow up with a Limbo case.

Reply:
Thanks @tetsuo-cpp!